File: 2023/AOC2023.Tests/Day02Tests.cs
using System.Text.RegularExpressions;
using Shouldly;

namespace AOC2023.Tests;

public class Day02Tests : TestBase
{
    [Test]
    public override void Part1_Sample()
    {
        // Arrange
        var games = get_sample()
            .Select(GetGameStats).ToList();
        var expectedSum = 8;
        
        // Act
        var possibleGames = games
            .Where(x => x.MaxRed <= _maxRed && x.MaxGreen <= _maxGreen && x.MaxBlue <= _maxBlue).ToList();
        var actualSum = possibleGames.Sum(x => x.GameNumber);

        // Assert
        actualSum.ShouldBe(expectedSum);
    }
    
    
    [Test]
    public override void Part1_Actual()
    {
        // Arrange
        var games = get_input()
            .Select(GetGameStats).ToList();
        var expectedSum = 2283;
        
        // Act
        var possibleGames = games
            .Where(x => x.MaxRed <= _maxRed && x.MaxGreen <= _maxGreen && x.MaxBlue <= _maxBlue).ToList();
        var actualSum = possibleGames.Sum(x => x.GameNumber);

        // Assert
        actualSum.ShouldBe(expectedSum);
    }
    
    [Test]
    public override void Part2_Sample()
    {
        // Arrange
        var games = get_sample()
            .Select(GetGameStats).ToList();
        var expectedSum = 2286;
        
        // Act
        var actualSum = games.Sum(x => x.MinPower);

        // Assert
        actualSum.ShouldBe(expectedSum);
    }

    
    [Test]
    public override void Part2_Actual()
    {
        // Arrange
        var games = get_input()
            .Select(GetGameStats).ToList();
        var expectedSum = 78669;
        
        // Act
        var actualSum = games.Sum(x => x.MinPower);

        // Assert
        actualSum.ShouldBe(expectedSum);
    }
    
    private int _maxRed = 12;
    private int _maxGreen = 13;
    private int _maxBlue = 14;

    private Game GetGameStats(string gameInput)
    {
        var gameRegex = new Regex(@"Game (?<gameNumber>\d+):(?<handfuls>.*)");
        var cubesRegex = new Regex(@"(?<cubes>\d+) (?<color>(blue|green|red))");
        var match = gameRegex.Match(gameInput);
        var game = new Game
        {
            GameNumber = int.Parse(match.Groups["gameNumber"].Value)
        };

        var handfuls = match.Groups["handfuls"].Value.Split(';');
        foreach (var handful in handfuls)
        {
            var cubeMatches = cubesRegex.Matches(handful);
            foreach (Match cubeMatch in cubeMatches)
            {
                switch (cubeMatch.Groups["color"].Value)
                {
                    case "red":
                        game.MaxRed = Math.Max(game.MaxRed, int.Parse(cubeMatch.Groups["cubes"].Value));
                        break;
                    case "green":
                        game.MaxGreen = Math.Max(game.MaxGreen, int.Parse(cubeMatch.Groups["cubes"].Value));
                        break;
                    case "blue":
                        game.MaxBlue = Math.Max(game.MaxBlue, int.Parse(cubeMatch.Groups["cubes"].Value));
                        break;
                }
            }
        }
        return game;
    }
    
    public class Game
    {
        public int GameNumber { get; set; }
        public int MaxRed { get; set; }
        public int MaxGreen { get; set; }
        public int MaxBlue { get; set; }
        public int MinPower => MaxRed * MaxBlue * MaxGreen;
    }
}

File: 2023/kotlin/aoc-2023/src/test/kotlin/Day02Tests.kt
import kotlin.math.max
import kotlin.test.assertEquals

class Day02Tests : TestBase() {
    override fun solvePart1Sample() {
        // Arrange
        val games = getSample()
            .map{getGameStats(it)}
        val expectedSum = 8

        // Act
        val possibleGames = games
            .filter { x -> x.MaxRed <= _maxRed && x.MaxGreen <= _maxGreen && x.MaxBlue <= _maxBlue }
        val actualSum = possibleGames.sumOf { it.gameNumber }

        // Assert
        assertEquals(expectedSum, actualSum)
    }

    override fun solvePart1Actual() {
        // Arrange
        val games = getInput()
            .map{getGameStats(it)}

        val expectedSum = 2283

        // Act
        val possibleGames = games
            .filter { x -> x.MaxRed <= _maxRed && x.MaxGreen <= _maxGreen && x.MaxBlue <= _maxBlue }
        val actualSum = possibleGames.sumOf { it.gameNumber }

        // Assert
        assertEquals(expectedSum, actualSum)
    }

    override fun solvePart2Sample() {
        // Arrange
        val games = getSample()
            .map{getGameStats(it)}
        val expectedSum = 2286

        // Act
        val actualSum = games.sumOf { it.MinPower }

        // Assert
        assertEquals(expectedSum, actualSum)
    }

    override fun solvePart2Actual() {
        // Arrange
        val games = getInput()
            .map{getGameStats(it)}
        val expectedSum = 78669

        // Act
        val actualSum = games.sumOf { it.MinPower }

        // Assert
        assertEquals(expectedSum, actualSum)
    }

    private val _maxRed: Int = 12
    private val _maxGreen: Int = 13
    private val _maxBlue: Int = 14

    private fun getGameStats(gameInput: String): Game{
        val gameRegex = Regex("Game (?<gameNumber>\\d+):(?<handfuls>.*)")
        val cubesRegex = Regex("(?<cubes>\\d+) (?<color>(blue|green|red))")
        val match = gameRegex.find(gameInput)
        val game = Game()
        game.gameNumber = match?.groups!!["gameNumber"]?.value?.toInt()!!

        val handfuls = match.groups["handfuls"]?.value?.split(';')?: listOf<String>()
        handfuls.forEach {
            handful ->
            val cubeMatches = cubesRegex.findAll(handful)
            cubeMatches.forEach {
                cubeMatch ->
                when (cubeMatch.groups["color"]?.value) {
                    "red" -> game.MaxRed = max(game.MaxRed, cubeMatch.groups["cubes"]?.value?.toInt()?:0)
                    "green" -> game.MaxGreen = max(game.MaxGreen, cubeMatch.groups["cubes"]?.value?.toInt()?:0)
                    "blue" -> game.MaxBlue = max(game.MaxBlue, cubeMatch.groups["cubes"]?.value?.toInt()?:0)
                }
            }
        }

        return game
    }

    class Game {
        var gameNumber: Int = 0
        var MaxRed: Int = 0
        var MaxGreen: Int = 0
        var MaxBlue: Int = 0
        val MinPower get() = MaxRed * MaxBlue * MaxGreen
    }
}